Driver logging: disable colors when logging to file

When debugging a large Miri trace, I like to put the trace into a file that I can analyze and search through in an editor. Unfortunately, since the recent logging changes, that leads to a lot of garbage in the file:
```
2:rustc│ │ │ │ │ │ ├─[2m0[0m[2mms[0m [1;32m INFO[0m [2mrustc_mir::interpret::step[0m Retag([fn entry] _1)
2:rustc│ │ │ │ │ │ ├─[2m0[0m[2mms[0m [1;32m INFO[0m [2mrustc_mir::interpret::step[0m StorageLive(_2)
2:rustc│ │ │ │ │ │ ├─[2m0[0m[2mms[0m [1;32m INFO[0m [2mrustc_mir::interpret::step[0m StorageLive(_3)
2:rustc│ │ │ │ │ │ ├─[2m0[0m[2mms[0m [1;32m INFO[0m [2mrustc_mir::interpret::step[0m _3 = move _1
2:rustc│ │ │ │ │ │ ├─[2m0[0m[2mms[0m [1;32m INFO[0m [2mrustc_mir::interpret::step[0m Retag(_3)
2:rustc│ │ │ │ │ │ ├─[2m0[0m[2mms[0m [1;32m INFO[0m [2mrustc_mir::interpret::step[0m _2 = std::mem::ManuallyDrop::<std::boxed::Box<T, A>>::new(move _3) -> [return: bb1, unwind: bb5]
2:rustc│ │ │ │ │ │ ├┐[2mrustc_mir::interpret::eval_context[0m::[1;32mframe[0m std::boxed::Box::<alloc::collections::btree::node::LeafNode<std::string::String, i32>>::into_unique
2:rustc│ │ │ │ │ │ │└┐[2mrustc_mir::interpret::eval_context[0m::[1;32mframe[0m std::mem::ManuallyDrop::<std::boxed::Box<alloc::collections::btree::node::LeafNode<std::string::String, i32>>>::new
2:rustc│ │ │ │ │ │ │ ├─[2m0[0m[2mms[0m [1;32m INFO[0m [2mrustc_mir::interpret::step[0m // executing bb0
2:rustc│ │ │ │ │ │ │ ├─[2m0[0m[2mms[0m [1;32m INFO[0m [2mrustc_mir::interpret::step[0m Retag([fn entry] _1)
2:rustc│ │ │ │ │ │ │ ├─[2m0[0m[2mms[0m [1;32m INFO[0m [2mrustc_mir::interpret::step[0m StorageLive(_2)
2:rustc│ │ │ │ │ │ │ ├─[2m0[0m[2mms[0m [1;32m INFO[0m [2mrustc_mir::interpret::step[0m _2 = move _1
2:rustc│ │ │ │ │ │ │ ├─[2m0[0m[2mms[0m [1;32m INFO[0m [2mrustc_mir::interpret::step[0m Retag(_2)
```
Looks like the color escape sequences are printed even when redirecting to a file. This makes working with the logfile much harder than it was before. Is there a way to fix that?

Cc @oli-obk

Hmm, I'm getting this error:

```
error[E0463]: can't find crate for `rustc_attr`
  --> src/lib.rs:14:1
   |
14 | extern crate rustc_attr;
   | ^^^^^^^^^^^^^^^^^^^^^^^^ can't find crate

error: aborting due to previous error
```

I'm using the latest master rustc and the latest miri. Is miri broken right now?

--- a/compiler/rustc_driver/src/lib.rs
+++ b/compiler/rustc_driver/src/lib.rs
@@ -1283,10 +1283,22 @@ pub fn init_env_logger(env: &str) {
         Ok(s) if s.is_empty() => return,
         Ok(_) => {}
     }
+    let color_logs = match std::env::var(String::from(env) + "_COLOR") {
+        Ok(value) => match value.as_ref() {
+            "always" => true,
+            "never" => false,
+            "auto" => stdout_isatty(),
+            _ => panic!("invalid log color value '{}': expected one of always, never, or auto", value),
+        },
+        Err(std::env::VarError::NotPresent) => stdout_isatty(),
+        Err(std::env::VarError::NotUnicode(_value)) => {
+            panic!("non-unicode log color value: expected one of always, never, or auto")
+        }
+    };
     let filter = tracing_subscriber::EnvFilter::from_env(env);
     let layer = tracing_tree::HierarchicalLayer::default()
         .with_indent_lines(true)
-        .with_ansi(stdout_isatty())
+        .with_ansi(color_logs)
         .with_targets(true)
         .with_thread_ids(true)
         .with_thread_names(true)

--- a/compiler/rustc_driver/src/lib.rs
+++ b/compiler/rustc_driver/src/lib.rs
@@ -1286,7 +1286,7 @@ pub fn init_env_logger(env: &str) {
     let filter = tracing_subscriber::EnvFilter::from_env(env);
     let layer = tracing_tree::HierarchicalLayer::default()
         .with_indent_lines(true)
-        .with_ansi(true)
+        .with_ansi(stdout_isatty())
         .with_targets(true)
         .with_thread_ids(true)
         .with_thread_names(true)

--- a/compiler/rustc_driver/src/lib.rs
+++ b/compiler/rustc_driver/src/lib.rs
@@ -1288,12 +1288,19 @@ pub fn init_env_logger(env: &str) {
             "always" => true,
             "never" => false,
             "auto" => stdout_isatty(),
-            _ => panic!("invalid log color value '{}': expected one of always, never, or auto", value),
+            _ => early_error(
+                ErrorOutputType::default(),
+                &format!(
+                    "invalid log color value '{}': expected one of always, never, or auto",
+                    value
+                ),
+            ),
         },
         Err(std::env::VarError::NotPresent) => stdout_isatty(),
-        Err(std::env::VarError::NotUnicode(_value)) => {
-            panic!("non-unicode log color value: expected one of always, never, or auto")
-        }
+        Err(std::env::VarError::NotUnicode(_value)) => early_error(
+            ErrorOutputType::default(),
+            "non-Unicode log color value: expected one of always, never, or auto",
+        ),
     };
     let filter = tracing_subscriber::EnvFilter::from_env(env);
     let layer = tracing_tree::HierarchicalLayer::default()
@@ -1324,7 +1331,7 @@ pub fn main() -> ! {
                 arg.into_string().unwrap_or_else(|arg| {
                     early_error(
                         ErrorOutputType::default(),
-                        &format!("Argument {} is not valid Unicode: {:?}", i, arg),
+                        &format!("argument {} is not valid Unicode: {:?}", i, arg),
                     )
                 })
             })

